doc/check-abi.1
.\" Automatically generated by Pod::Man 4.14 (Pod::Simple 3.42)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "CHECK-ABI 1"
.TH CHECK-ABI 1 "2022-03-07" "annobin-1" "RPM Development Tools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
check\-abi \- ABI Checking
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
check-abi
  [\fB\-\-help\fR]
  [\fB\-\-version\fR]
  [\fB\-\-verbose\fR]
  [\fB\-\-quiet\fR]
  [\fB\-\-silent\fR]
  [\fB\-\-inconsistencies\fR]
  [\fB\-\-ignore\-unknown\fR]
  [\fB\-\-ignore\-\fR\fIABI|enum|FORTIFY|stack\-prot\fR]
  [\fB\-\-readelf=\fR\fIpath\fR]
  [\fB\-\-tmpdir=\fR\fIdir\fR]
  [\fB\-\-\fR]
  \fIfile\fR...
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fIcheck-abi\fR script reports any potential \s-1ABI\s0 conflicts in
the files specified.  This includes the use of the
\&\fB\-fshort\-enums\fR option, the \fB\-fstack\-protector\fR option
and the \fB\-D_FORTIFY_SOURCE\fR option.  All of these can affect
passing data between functions and hence should be used uniformly
throughout the binary.
.PP
The script accepts the following command line options:
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
.PD 0
.IP "\fB\-h\fR" 4
.IX Item "-h"
.PD
Displays the usage of the script and then exits.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
.PD 0
.IP "\fB\-v\fR" 4
.IX Item "-v"
.PD
Displays the version of the script.
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
.PD 0
.IP "\fB\-V\fR" 4
.IX Item "-V"
.PD
Enables verbose mode, causing the script to detail each action it
takes.
.IP "\fB\-\-quiet\fR" 4
.IX Item "--quiet"
.PD 0
.IP "\fB\-q\fR" 4
.IX Item "-q"
.PD
Do not include the name of script in the out generated by the script.
.IP "\fB\-\-silent\fR" 4
.IX Item "--silent"
.PD 0
.IP "\fB\-s\fR" 4
.IX Item "-s"
.PD
Produce no output.  Just return an exit status.
.IP "\fB\-\-inconsistencies\fR" 4
.IX Item "--inconsistencies"
.PD 0
.IP "\fB\-i\fR" 4
.IX Item "-i"
.PD
Only report files with potential \s-1ABI\s0 problems.
.IP "\fB\-\-ignore\-unknown\fR" 4
.IX Item "--ignore-unknown"
Do not report file types that are not supported or recognised.
.IP "\fB\-\-ignore\-\fR\fIABI|enum|FORTIFY|stack\-prot\fR" 4
.IX Item "--ignore-ABI|enum|FORTIFY|stack-prot"
Disables individual \s-1ABI\s0 checks.  Multiple occurrences of this option
accumulate.  Possible option values are:
.RS 4
.IP "\fB\s-1ABI\s0\fR" 4
.IX Item "ABI"
Disable checks of the general \s-1ABI\s0 information.
.IP "\fBenum\fR" 4
.IX Item "enum"
Disable checks of the \fB\-fshort\-enum\fR option.
.IP "\fB\s-1FORTIFY\s0\fR" 4
.IX Item "FORTIFY"
Disable checks of the \fB\-D_FORTIFY_SOURCE\fR option.
.IP "\fBstack-prot\fR" 4
.IX Item "stack-prot"
Disable checks of the \fB\-fstack\-protect\fR option.
.RE
.RS 4
.RE
.IP "\fB\-\-tmpdir=\f(BIdir\fB\fR" 4
.IX Item "--tmpdir=dir"
.PD 0
.IP "\fB\-t=\f(BIdir\fB\fR" 4
.IX Item "-t=dir"
.PD
Directory to use to store temporary files.
.IP "\fB\-\-readelf=\f(BIpath\fB\fR" 4
.IX Item "--readelf=path"
.PD 0
.IP "\fB\-r=\f(BIpath\fB\fR" 4
.IX Item "-r=path"
.PD
Use the specified program to read the notes from the files.
.IP "\fB\-\-\fR" 4
.IX Item "--"
Stop accumulating command line options.  This allows the script to be
run on files whose names starts with a dash.
.SH "OPTIONS"
.IX Header "OPTIONS"
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2018 \- 2022 Red Hat.
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the \s-1GNU\s0 Free Documentation License, Version 1.3
or any later version published by the Free Software Foundation;
with no Invariant Sections, with no Front-Cover Texts, and with no
Back-Cover Texts.  A copy of the license is included in the
section entitled \*(L"\s-1GNU\s0 Free Documentation License\*(R".


scripts/check-abi.sh
#!/bin/bash

# Script to check for ABI conflicts in annotated binaries.
#
# Created by Nick Clifton.  <nickc@redhat.com>
# Copyright (c) 2017-2018 Red Hat.
#
# This is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published
# by the Free Software Foundation; either version 3, or (at your
# option) any later version.

# It is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# Usage:
#   check-abi.sh [switches] file(s)
#
# This script does not handle directories.  This is deliberate.
# It is intended that if recursion is needed then it will be
# invoked from find, like this:
#
#   find . -type f -exec check-abi.sh {} \;

version=3.2

help ()
{
  # The following exec goop is so that we don't have to manually
  # redirect every message to stderr in this function.
  exec 4>&1    # save stdout fd to fd #4
  exec 1>&2    # redirect stdout to stderr

  cat <<__EOM__

This is a shell script to check that the given program(s) have not
been built with object files that contain conflicting ABI options.

Usage: $prog {files|options}

  {options} are:
  -h        --help             Display this information.
  -v        --version          Report the version number of this script.
  -s        --silent           Produce no output, just an exit status.
  -q        --quiet            Do not include the script name in the output.
  -V        --verbose          Report on progress.
  -i        --inconsistencies  Only report potential ABI problems.
  -r=<PATH> --readelf=<PATH>   Path to version of readelf to use to read notes.
  -t=<PATH> --tmpfile=<PATH>   Temporary file to use.

  --ignore-unknown             Silently skip files of unknown type.
  --ignore-ABI                 Do not check ABI annotation.
  --no-ignore-ABI              Check ABI information but do not complain if none is found.
  --ignore-enum                Do not check enum size annotation.
  --no-ignore-enum             Check enum size information but do not complain if none is found.
  --ignore-FORTIFY             Do not check FORTIFY SOURCE annotation.
  --no-ignore-FORTIFY          Check FORTIFY SOURCE information but do not complain if none is found.
  --ignore-stack-prot          Do not check stack protection annotation.
  --no-ignore-stack-prot       Check stack protection information but do not complain if none is found.
  --ignore-gaps                Do not fail if there are gaps in the coverage.
  --no-ignore-gaps             Fail if there are gaps in the coverage.

  --                           Stop accumulating options.

__EOM__
  exec 1>&4   # Copy stdout fd back from temporary save fd, #4
}

main ()
{
    init
    
    parse_args ${1+"$@"}

    scan_files

    if [ $failed -ne 0 ];
    then
	exit 1
    else
	exit 0
    fi
}

report ()
{
    if [ $silent -ne 0 ];
    then
	return
    fi

    if [ $quiet -eq 0 ];
    then
	echo -n $prog": "
    fi
    
    echo ${1+"$@"}
}

fail ()
{
    report "Internal error: " ${1+"$@"}
    exit 1
}

verbose ()
{
    if [ $verb -ne 0 ]
    then
	report ${1+"$@"}
    fi
}

# Initialise global variables.
init ()
{
    files[0]="";  
    # num_files is the number of files to be listed minus one.
    # This is because we are indexing the files[] array from zero.
    num_files=0;

    failed=0
    silent=0
    verb=0
    quiet=0
    inconsistencies=0
    ignore_abi=0
    ignore_enum=0
    ignore_fortify=0
    ignore_stack_prot=0
    ignore_unknown=0
    ignore_gaps=0
    scanner=readelf
    tmpfile=/dev/shm/check.abi.delme
}

# Parse our command line
parse_args ()
{
    prog=`basename $0`;

    # Locate any additional command line switches
    # Likewise accumulate non-switches to the files list.
    while [ $# -gt 0 ]
    do
	optname="`echo $1 | sed 's,=.*,,'`"
	optarg="`echo $1 | sed 's,^[^=]*=,,'`"

	case "$optname" in
	    -v | --version)
		report "version: $version"
		exit 0
		;;
	    -h | --help)
		help 
		exit 0
		;;
	    -s | --silent)
		silent=1;
		verb=0;
		;;
	    -q | --quiet)
		quiet=1;
		;;
	    -V | --verbose)
		silent=0;
		verb=1;
		;;
	    -i | --inconsistencies)
		silent=0;
		inconsistencies=1;
		;;
	    -r | --readelf)
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "-$optname option needs a program name"
		    else
			scanner=$1
		    fi
		else
		    scanner="$optarg"
		fi
		;;
	    -t | --tmpfile)
		if test "x$optarg" = "x$optname" ;
		then
		    shift
		    if [ $# -eq 0 ]
		    then
			fail "-t option needs a file name"
		    else
			tmpfile=$1
		    fi
		else
		    tmpfile="$optarg"
		fi
		;;

	    --ignore-unknown)
		ignore_unknown=1;
		;;
	    --ignore-abi | --ignore-ABI)
		ignore_abi=1;
		;;
	    --no-ignore-abi | --no-ignore-ABI)
		ignore_abi=2;
		;;
	    --ignore-enum)
		ignore_enum=1;
		;;
	    --no-ignore-enum)
		ignore_enum=2;
		;;
	    --ignore-fortify | --ignore-FORTIFY)
		ignore_fortify=1;
		;;
	    --no-ignore-fortify | --no-ignore-FORTIFY)
		ignore_fortify=2;
		;;
	    --ignore-stack-prot)
		ignore_stack_prot=1;
		;;
	    --no-ignore-stack-prot)
		ignore_stack_prot=2;
		;;
            --ignore-gaps)
                ignore_gaps=1;
                ;;
            --no-ignore-gaps)
                ignore_gaps=0;
                ;;
	    
	    --)
		shift
		break;
		;;
	    --*)
		report "unrecognised option: $1"
		help
		exit 1
		;;
	    *)
		files[$num_files]="$1";
		let "num_files++"
		;;
	esac
	shift
    done

    # Accumulate any remaining arguments without processing them.
    while [ $# -gt 0 ]
    do
	files[$num_files]="$1";
	let "num_files++";
	shift
    done

    if [ $num_files -gt 0 ];
    then
	# Remember that we are counting from zero not one.
	let "num_files--"
    else
	report "must specify at least one file to scan"
	exit 1
    fi
}

scan_files ()
{
    local i

    i=0;
    while [ $i -le $num_files ]
    do
	scan_file i
	let "i++"
    done
}

scan_file ()
{
    local file

    # Paranoia checks - the user should never encounter these.
    if test "x$1" = "x" ;
    then
	fail "scan_file called without an argument"
    fi
    if test "x$2" != "x" ;
    then
	fail "scan_file called with too many arguments"
    fi

    # Use quotes when accessing files in order to preserve
    # any spaces that might be in the directory name.
    file="${files[$1]}";

    # Catch names that start with a dash - they might confuse readelf
    if test "x${file:0:1}" = "x-" ;
    then
	file="./$file"
    fi

    if ! [ -a "$file" ]
    then
	if [ $ignore_unknown -eq 0 ];
	then
	    report "$file: file not found"
	    failed=1
	fi
	return
    fi
    
    if ! [ -f "$file" ]
    then
	if [ $ignore_unknown -eq 0 ];
	then
	    report "$file: not an ordinary file"
	    failed=1
	fi
	return
    fi

    if ! [ -r "$file" ]
    then
	if [ $ignore_unknown -eq 0 ];
	then
	    report "$file: not readable"
	    failed=1
	fi
	return
    fi

    file $file | grep --silent -e ELF
    if [ $? != 0 ];
    then
	if [ $ignore_unknown -eq 0 ];
	then
	    report "$file: not an ELF format file"
	    failed=1
	fi
	return
    fi
    
    $scanner --wide --notes $file > $tmpfile 2>&1
    if [ $? != 0 ];
    then
	report "$file: scanner '$scanner' failed - see $tmpfile"
	failed=1
	# Leave the tmpfile intact so that it can be examined by the user.
	return
    fi

    grep -q -e "Unknown note" $tmpfile
    if [ $? == 0 ];
    then
	# The fortify and stack protection checks need parsed notes.
	if [[ $ignore_fortify -eq 0 || $ignore_stack_prot -eq 0 ]];
	then
	    report "$file: scanner '$scanner' could not parse the notes - see $tmpfile"
	    failed=1
	    # Leave the tmpfile intact so that it can be examined by the user.
	    return
	fi
    fi

    verbose "NOTES: `cat $tmpfile`"

    if [ $ignore_gaps -eq 0 ];
    then
        grep -q -e "Gap in build notes" $tmpfile
        if [ $? == 0 ];
        then
	    report "$file: there are gaps in the build notes"
	    failed=1
        fi
    fi       

    local -a abis

    if [ $ignore_abi -ne 1 ];
    then
	# Convert:
	#   *<ABI>0x145e82c442000192 0x00000000 NT_GNU_BUILD...
	# or:
	#   GA*<ABI>0x145e82c442000192 0x00000000 NT_GNU_BUILD...
	# into:
	#   abis[n]=145e82c442000192

	eval 'abis=($(grep -e \<ABI\> $tmpfile | cut -d " " -f 3 | cut -d x -f 2 | sort -u))'

	verbose "ABI Info: ${abis[*]}"

	if [ ${#abis[*]} -lt 1 ];
	then
	    if [[ $ignore_abi -eq 0 && $inconsistencies -eq 0 ]];
	    then
		report "$file: does not have an ABI note"
	    fi
	else
	    if [ ${#abis[*]} -gt 1 ];
	    then
		local i mismatch=0

		if [ $inconsistencies -eq 0 ];
		then
		    report "$file: contains ${#abis[*]} ABI notes"
		fi

		i=1;
		while [ $i -lt ${#abis[*]} ]
		do
		    if test "${abis[i]}" != "${abis[i-1]}" ;
		    then
			# FIXME: Add code to differentiate between functions which have changed ABI and files ?
			report "$file: differing ABI values detected: ${abis[i]} vs ${abis[i-1]}"
			failed=1
			mismatch=1
		    fi
		    let "i++"
		done

		if [ $mismatch -eq 0 ];
		then
		    verbose "$file: ABI: ${abis[0]}"
		fi
	    fi
	fi
    fi

    if [ $ignore_enum -ne 1 ];
    then
	# Convert:
	#   +<short enum>true or  GA+<short enum>true
	# into:
	#   abis[n]=true
	# and
	#   !<short enum>false or  GA!<short enum>false
	# into:
	#   abis[n]=false

	eval 'abis=($(grep -e "short enum" $tmpfile | cut -f 2 -d ">" | cut -f 1 -d " " | sort -u))'

	verbose "Enum Info: ${abis[*]}"

	if [ ${#abis[*]} -lt 1 ];
	then
	    if [[ $ignore_enum -eq 0 && $inconsistencies -eq 0 ]];
	    then
		report "$file: does not record enum size"
	    fi
	else
	    if [ ${#abis[*]} -gt 1 ];
	    then
		local i mismatch=0

		if [ $inconsistencies -eq 0 ];
		then
		    report "$file: contains ${#abis[*]} enum size notes"
		fi
		i=1;
		while [ $i -lt ${#abis[*]} ]
		do
		    if test "${abis[i]}" != "${abis[i-1]}" ;
		    then
			report "$file: differing -fshort-enums detected: ${abis[i]} vs ${abis[i-1]}"
			failed=1
			mismatch=1
		    fi
		    let "i++"
		done

		if [ $mismatch -eq 0 ];
		then
		    verbose "$file: -fshort-enums: ${abis[0]}"
		fi
	    fi
	fi
    fi

    if [ $ignore_fortify -ne 1 ];
    then
	# Convert:
	#   *FORTIFY:0x1
	# or:
	#   GA*FORTIFY:0x1
	# into:
	#   abis[n]=1

	eval 'abis=($(grep -e FORTIFY $tmpfile | cut -f 2 -d ":" | cut -b 3-5 | sed -e "s/ff/unknown/" | sort -u))'

	verbose "Fortify Info: ${abis[*]}"

	if [ ${#abis[*]} -lt 1 ];
	then
	    if [[ $ignore_fortify -eq 0 && $inconsistencies -eq 0 ]];
	    then
		report "$file: does not record _FORTIFY_SOURCE level"
	    fi
	else
	    if [ ${#abis[*]} -gt 1 ];
	    then
		local i mismatch=0

		if [ $inconsistencies -eq 0 ];
		then
		    report "$file: contains ${#abis[*]} FORTIFY_SOURCE notes"
		fi
		i=1;
		while [ $i -lt ${#abis[*]} ]
		do
		    if test "${abis[i]}" != "${abis[i-1]}" ;
		    then
			report "$file: differing FORTIFY SOURCE levels: ${abis[i]} vs ${abis[i-1]}"
			failed=1;
			mismatch=1;
		    fi
		    let "i++"
		done

		if [ $mismatch -eq 0 ];
		then
		    verbose "$file: -D_FORTIFY_SOURCE=${abis[0]}"
		fi
	    fi
	fi
    fi

    if [ $ignore_stack_prot -ne 1 ];
    then
	# Convert:
	#   *<stack prot><type>
	# into:
	#   abis[n]=<type>

	eval 'abis=($(grep -e "stack prot" $tmpfile | cut -f 4 -d " " | cut -b 6- | sort -u))'

	verbose "Stack Protection Info: ${abis[*]}"

	if [ ${#abis[*]} -lt 1 ];
	then
	    if [[ $ignore_stack_prot -eq 0 && $inconsistencies -eq 0 ]];
	    then
		report "$file: does not record -fstack-protect status"
	    fi
	else
	    if [ ${#abis[*]} -eq 1 ];
	    then
		verbose "$file: -fstack-protect=${abis[0]}"
	    else
		local i mismatch=0

		if [ $inconsistencies -eq 0 ];
		then
		    report "$file: contains ${#abis[*]} stack protection notes"
		fi
		i=1;
		while [ $i -lt ${#abis[*]} ]
		do
		    if test "${abis[i]}" != "${abis[i-1]}" ;
		    then
			report "$file: differing stack protection levels: ${abis[i]} vs ${abis[i-1]}"
			failed=1;
			mismatch=1;
		    fi
		    let "i++"
		done

		if [ $mismatch -eq 0 ];
		then
		    verbose "$file: -fstack-protect=${abis[0]}"
		fi
	    fi
	fi
    fi

    rm -f $tmpfile
}

# Invoke main
main ${1+"$@"}
